src/forms/Y2021/irsForms/F8582.ts

```diff
@@ -2,8 +2,14 @@ import { TaxPayer } from 'ustaxes/core/data'
 import ScheduleE, { MatrixRow } from './ScheduleE'
 import log from 'ustaxes/core/log'
 
+let suppressedLogMessages: string[] = []
+
 const unimplemented = (message: string): void => {
-  log.warn(`[Form 8582]: ${message}`)
+  // Excessive logging can hang the dev tools, so limit messages
```

Comment:
I've been working a bit on memoize.  I want to try some stuff after this gets merged.  I'm thinking that I could have every line look something like this:

```typescript
l19c = (): number | undefined => roundAndMemoize("l19c", ...)
```

By doing that, I can:
1.  Have it round each line if the user selects that option and
2. Memoize the value

Reply:
still think that rounding should be saved until later. There are cases where it matters. K-1s, 8949 can have arbitrary numbers of fields that need to be added together, so the rounding error can grow to roughly $(n/2). It is annoying though to see the off by one errors on the PDFs though i get it.

Reply:
I think it depends on the user's goal.  Mine is to use this to verify my work before using https://www.freefilefillableforms.com to actually file.  When you enter any number there, they immediately round it.  So I'd like to add an option to match that if you're okay with that.  I wouldn't make it the default.

Reply:
That makes sense then, wrapping all the line methods with something that memoizes and applies one of the two apparently acceptable rounding strategies could work.

Let's handle it in another issue/PR. I'm trying to think if there's a way to avoid having to remember to add this behavior to every field in the project. It would be nice if we could inject the behavior somehow.

Reply:
Here's one way I was thinking of doing this.

**Update `Form` with methods to set up memoized calculations**
![image](https://user-images.githubusercontent.com/1426848/156905773-70af38b8-b062-41ef-9e81-3659f3a7d8ca.png)

**Update all lines in all forms to use `lineFn` or `lineFnNumber`**
![image](https://user-images.githubusercontent.com/1426848/156905795-3294c57a-9fe7-4127-96a4-89a5dbf66b51.png)

What do you think?  I'd need to figure out where to put the setting for rounding, but that would be built into `lineFn` and `lineFnNumber`.

In a test on 1040 line 11, I think it used the memorized version around 485 times.

<hr/>

I was also looking into why it's quite slow to generate the PDF.  I found the following.

**This line takes about 1 second:**
https://github.com/ustaxes/UsTaxes/blob/b815f743bd0ca02a9d60a59f9e5a648594607334/src/forms/YearForms.ts#L66
**This line takes about 6 seconds:**
https://github.com/ustaxes/UsTaxes/blob/b815f743bd0ca02a9d60a59f9e5a648594607334/src/forms/YearForms.ts#L67

Any thoughts on how to solve that?  Do you think it's worth trying to speed up or do you think it should just be masked with a loading overlay?

Reply:
@zakpatterson, do you have thoughts on this?

Reply:
Let's make a new issue to investigate the slowness here, I do think we need to change strategy with it a bit. 

combinePDFs does a lot of work, and rendering PDFs is pretty slow for sure. But it looks like there could be some duplication of effort in that implementation. 

Keep in mind that we have form data on numerous attachments that have to be combined into one attachment. I found a while ago that the only way to do this is to reserialize the PDFDocuments so that the form data is frozen into the forms. As a result the outputted PDF is no longer editable.

So essentially each document is being serialized, then every page is being copied over one at a time into a new document. But I don't know why this should take so long.